database.go

```diff
@@ -0,0 +1,723 @@
+	for _, rp := range db.policies {
+		o.Policies = append(o.Policies, rp)
+	}
+	for _, s := range db.shards {
+		o.Shards = append(o.Shards, s)
+	}
+	return json.Marshal(&o)
+}
+
+// UnmarshalJSON decodes a JSON-encoded byte slice to a database.
+func (db *database) UnmarshalJSON(data []byte) error {
+	// Decode into intermediate type.
+	var o databaseJSON
+	if err := json.Unmarshal(data, &o); err != nil {
+		return err
+	}
+
+	// Copy over properties from intermediate type.
+	db.name = o.Name
+	db.defaultRetentionPolicy = o.DefaultRetentionPolicy
+
+	// Copy shard policies.
+	db.policies = make(map[string]*RetentionPolicy)
+	for _, rp := range o.Policies {
+		db.policies[rp.Name] = rp
+	}
+
+	// Copy shards.
+	db.shards = make(map[uint64]*Shard)
+	for _, s := range o.Shards {
+		db.shards[s.ID] = s
+	}
+
+	return nil
+}
+
+// databaseJSON represents the JSON-serialization format for a database.
+type databaseJSON struct {
+	Name                   string             `json:"name,omitempty"`
+	DefaultRetentionPolicy string             `json:"defaultRetentionPolicy,omitempty"`
+	Policies               []*RetentionPolicy `json:"policies,omitempty"`
+	Shards                 []*Shard           `json:"shards,omitempty"`
+}
+
+// Measurement represents a collection of time series in a database. It also contains in memory
+// structures for indexing tags. These structures are accessed through private methods on the Measurement
+// object. Generally these methods are only accessed from Index, which is responsible for ensuring
+// go routine safe access.
+type Measurement struct {
+	Name   string    `json:"name,omitempty"`
+	Fields []*Fields `json:"fields,omitempty"`
+
+	// in memory index fields
+	series              map[string]*Series // sorted tagset string to the series object
+	seriesByID          map[uint32]*Series // lookup table for series by their id
+	measurement         *Measurement
+	seriesByTagKeyValue map[string]map[string]SeriesIDs // map from tag key to value to sorted set of series ids
+	ids                 SeriesIDs                       // sorted list of series IDs in this measurement
+}
+
+func NewMeasurement(name string) *Measurement {
+	return &Measurement{
+		Name:   name,
+		Fields: make([]*Fields, 0),
+
+		series:              make(map[string]*Series),
+		seriesByID:          make(map[uint32]*Series),
+		seriesByTagKeyValue: make(map[string]map[string]SeriesIDs),
+		ids:                 SeriesIDs(make([]uint32, 0)),
+	}
+}
+
+// addSeries will add a series to the measurementIndex. Returns false if already present
+func (m *Measurement) addSeries(s *Series) bool {
+	if _, ok := m.seriesByID[s.ID]; ok {
+		return false
+	}
+	m.seriesByID[s.ID] = s
+	tagset := string(marshalTags(s.Tags))
+	m.series[tagset] = s
+	m.ids = append(m.ids, s.ID)
+	// the series ID should always be higher than all others because it's a new
+	// series. So don't do the sort if we don't have to.
+	if len(m.ids) > 1 && m.ids[len(m.ids)-1] < m.ids[len(m.ids)-2] {
```

Comment:
I see. As you have indicated, the requires that `m` always increment, but you have no control over that here -- you expect it to be true, so everything should be fine, but you're protecting yourself against the design changing in the future.

My biggest concern right now is that it is a little confusing to the next guy. If this sorting takes place, I'd log a warn-level message perhaps. After all this shouldn't be necessary. Alternatively panic or assert.